Sources/Sentry/SentrySystemEventsBreadcrumbs.m

```diff
@@ -0,0 +1,174 @@
+#import "SentrySystemEventsBreadcrumbs.h"
+#import "SentrySDK.h"
+#import "SentryLog.h"
+
+// all those notifications are not available for tvOS
+#if TARGET_OS_IOS
+#    import <UIKit/UIKit.h>
+#endif
+
+@implementation SentrySystemEventsBreadcrumbs
+
+- (void)start
+{
+#if TARGET_OS_IOS
+    UIDevice *currentDevice = [UIDevice currentDevice];
+    if (currentDevice == nil) {
+        [SentryLog logWithMessage:@"currentDevice is null, it won't be able to record battery breadcrumbs."
+                         andLevel:kSentryLogLevelDebug];
+        return;
+    }
+    
+    [self initBatteryObserver:currentDevice];
+    [self initOrientationObserver:currentDevice];
+    [self initKeyboardVisibilityObserver];
+    [self initScreenshotObserver];
+    
+#else
+    [SentryLog logWithMessage:@"NO UIKit -> [SentrySystemEventsBreadcrumbs.start] does nothing."
+                     andLevel:kSentryLogLevelDebug];
+#endif
+}
+
+#if TARGET_OS_IOS
+- (void)initBatteryObserver:(UIDevice*)currentDevice
+{
+    if (currentDevice.batteryMonitoringEnabled == NO) {
+        currentDevice.batteryMonitoringEnabled = YES;
+    }
+    
+    NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
+    
+    // https://developer.apple.com/documentation/uikit/uidevicebatteryleveldidchangenotification
+    [defaultCenter addObserver:self selector:@selector(batteryStateChanged:) name:UIDeviceBatteryLevelDidChangeNotification object:currentDevice];
+    // https://developer.apple.com/documentation/uikit/uidevicebatterystatedidchangenotification
+    [defaultCenter addObserver:self selector:@selector(batteryStateChanged:) name:UIDeviceBatteryStateDidChangeNotification object:currentDevice];
+    
+    // for testing only
+    // [[NSNotificationCenter defaultCenter] postNotificationName:@"UIDeviceBatteryStateDidChangeNotification" object:currentDevice];
+}
+#endif
+
+#if TARGET_OS_IOS
+- (void)batteryStateChanged:(NSNotification*)notification
+{
+    // Notifications for battery level change are sent no more frequently than once per minute
+    NSDictionary *batteryData = [self getBatteryStatus:notification.object];
```

Comment:
`m`: Why don't we call `[UIDevice currentDevice]` here instead of getting it via the notification?

Reply:
why would we call `[UIDevice currentDevice]` if the param is available thru the `notification`?

Reply:
Now that I think about it again, I would keep it. It is going to make it easier to test.